compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala

```diff
@@ -381,10 +382,23 @@ class TreePickler(pickler: TastyPickler) {
                 pickleType(tp)
               }
             case _ =>
-              writeByte(if (name.isTypeName) SELECTtpt else SELECT)
               val sig = tree.tpe.signature
-              pickleNameAndSig(name, sig)
-              pickleTree(qual)
+              val isAmbiguous =
+                sig != Signature.NotAMethod
+                && qual.tpe.nonPrivateMember(name).match
+                  case d: MultiDenotation => d.atSignature(sig).isInstanceOf[MultiDenotation]
+                  case _ => false
+              if isAmbiguous then
+                writeByte(SELECTin)
+                withLength {
+                  pickleNameAndSig(name, tree.symbol.signature)
+                  pickleTree(qual)
+                  pickleType(tree.symbol.owner.typeRef)
```

Comment:
I don't understand how this is enough for disambiguation. In #9050 both overloads of `append` are defined in the same owner: `Buffer`. Shouldn't we pickle the type of the overload itself?

Reply:
The signature of the symbol is enough. The two signatures are always different or we would get a "have the same erasure" error. But it has to be the signature of the symbol itself, not the signature after asSeenFrom.

Reply:
Ah I see, I missed that the signature comes from the symbol here. I think this is worth explaining in more details in the TastyFormat documentation since it's pretty subtle and doing the wrong thing will appear to work most of the time.